🏰 [vllm] Support `base_url` parameter for vLLM client initialization

re-imagined opened the pull request:
related issue https://github.com/huggingface/trl/issues/3322
- Introduce base_url parameter to simplify server connection
- Update __init__ method to support both base_url and host+port configurations
- Modify URL construction in various methods to use base_url
- Update documentation to include new initialization examples

Close #3322.

qgallouedec commented on tests/test_vllm_client_server.py:
```diff
@@ -162,3 +238,64 @@ def tearDownClass(cls):
+        # Initialize the client with base_url
+        cls.client = VLLMClient(base_url="http://localhost:8000", connection_timeout=120)
+
+    def test_generate(self):
+        prompts = ["Hello, AI!", "Tell me a joke"]
+        outputs = self.client.generate(prompts)
+
+        # Check that the output is a list
+        self.assertIsInstance(outputs, list)
+
+        # Check that the number of generated sequences is equal to the number of prompts
+        self.assertEqual(len(outputs), len(prompts))
+
+        # Check that the generated sequences are lists of integers
+        for seq in outputs:
+            self.assertTrue(all(isinstance(tok, int) for tok in seq))
+
+    def test_update_model_params(self):
+        model = AutoModelForCausalLM.from_pretrained(self.model_id, device_map="cuda")
+        self.client.update_model_params(model)
+
+    def test_reset_prefix_cache(self):
+        # Test resetting the prefix cache
+        self.client.reset_prefix_cache()
+
+    @classmethod
+    def tearDownClass(cls):
+        super().tearDownClass()
+
+        # Close the client
+        cls.client.close_communicator()
+
+        # vLLM x pytest (or Popen) seems not to handle process termination well. To avoid zombie processes, we need to
+        # kill the server process and its children explicitly.
+        parent = psutil.Process(cls.server_process.pid)
+        children = parent.children(recursive=True)
+        for child in children:
+            child.send_signal(signal.SIGTERM)
+        cls.server_process.terminate()
+        cls.server_process.wait()
```
```suggestion
```

I think one test is enough

qgallouedec commented on trl/extras/vllm_client.py:
```diff
@@ -47,10 +45,12 @@ class VLLMClient:
 
     Args:
+        base_url (`str`, *optional*, defaults to `None`):
+            Base URL for the vLLM server (e.g., "http://localhost:8000"). If provided, host and server_port are ignored.
```
```suggestion
            Base URL for the vLLM server (e.g., `"http://localhost:8000"`). If provided, host and server_port are ignored.
```

qgallouedec commented on trl/extras/vllm_client.py:
```diff
@@ -47,10 +45,12 @@ class VLLMClient:
+            Base URL for the vLLM server (e.g., "http://localhost:8000"). If provided, host and server_port are ignored.
         host (`str`, *optional*, defaults to `"0.0.0.0"`):
-            IP address of the vLLM server.
+            IP address of the vLLM server. Ignored if base_url is provided.
```
```suggestion
            IP address of the vLLM server. Ignored if `base_url` is provided.
```

qgallouedec commented on trl/extras/vllm_client.py:
```diff
@@ -47,10 +45,12 @@ class VLLMClient:
+            IP address of the vLLM server. Ignored if base_url is provided.
         server_port (`int`, *optional*, defaults to `8000`):
-            Port number of the vLLM server.
+            Port number of the vLLM server. Ignored if base_url is provided.
```
```suggestion
            Port number of the vLLM server. Ignored if `base_url` is provided.
```

qgallouedec commented on trl/extras/vllm_client.py:
```diff
@@ -83,16 +101,30 @@ class VLLMClient:
+        if base_url is not None:
+            # Parse the base_url to extract host and port
+            parsed_url = urlparse(base_url)
+            scheme = parsed_url.scheme or "http"
```
this allows to add something like `"localhost:8000"` instead of `"http://localhost:8000"` right?

re-imagined replied:
yes

qgallouedec commented on trl/trainer/grpo_trainer.py:
```diff
@@ -627,9 +627,16 @@ def data_collator(features):  # No data collation is needed in GRPO
-                self.vllm_client = VLLMClient(
-                    args.vllm_server_host, args.vllm_server_port, connection_timeout=args.vllm_server_timeout
-                )
+                if args.vllm_server_base_url is not None:
+                    self.vllm_client = VLLMClient(
+                        base_url=args.vllm_server_base_url, connection_timeout=args.vllm_server_timeout
+                    )
+                else:
+                    self.vllm_client = VLLMClient(
+                        host=args.vllm_server_host, 
+                        server_port=args.vllm_server_port, 
+                        connection_timeout=args.vllm_server_timeout
+                    )
```
maybe this is easier?

```suggestion
                base_url = args.vllm_server_base_url if args.vllm_server_base_url is not None else f"http://{args.vllm_server_port}:{args.vllm_server_port}"
                self.vllm_client = VLLMClient(base_url=ase_url, connection_timeout=args.vllm_server_timeout)
```

qgallouedec reviewed (commented):
I think this change makes sense, thanks for suggesting it.
Sorry for the delayed review.
I've made a few comments.
Have you tested on your infrastructure?

re-imagined committed:
refactor

re-imagined commented on trl/extras/vllm_client.py:
```diff
@@ -207,14 +239,14 @@ def init_communicator(self):
 
         # Set up the communication group for weight broadcasting
-        pg = StatelessProcessGroup.create(host=self.host, port=self.group_port, rank=self.rank, world_size=world_size)
+        pg = StatelessProcessGroup.create(host="0.0.0.0", port=self.group_port, rank=self.rank, world_size=world_size)
```
yeah, when host is not provided at initialization of the client, using `self.host` to create StatelessProcessGroup could cause error

re-imagined commented:
> I think this change makes sense, thanks for suggesting it. Sorry for the delayed review. I've made a few comments. Have you tested on your infrastructure?

Thanks for your review. 😄
I am trying to test it on different cases, and checking if the model output is correct, but I am struggling on some nccl error when executing the tests. (it worked well on last commit 🤣, until I merge the main branch)
I will let you know as soon as i have an update

re-imagined committed:
refactor

re-imagined committed:
refactor

re-imagined committed:
refactor

re-imagined committed:
Merge branch 'main' into vllm_client_custom_url

re-imagined committed:
fix

qgallouedec commented:
Hey! any update?

re-imagined committed:
fix host

re-imagined commented:
> Hey! any update?

hi @qgallouedec,
Sorry for late update, I've been out for few days.

After several tests, I realized that  using 0.0.0.0 as the host address in StatelessProcessGroup.create isn't viable—it requires the server's actual IP address to function properly. But if we use base url to init the vllm client, we don't know the server's ip. 

Proposed Approach:  Use the base url to resolve the server's IP programmatically in the client  (when calling `/health`).

I’m currently testing this approach and diving deeper into how StatelessProcessGroup interacts with PyTorch’s TCPStore

I’ll update this thread once I’ve narrowed down the root cause and validated the fix. Let me know if you have insights or suggestions!

re-imagined committed:
Merge branch 'main' into vllm_client_custom_url

re-imagined committed:
fix(vllm_client): update host value using X-Forwarded-For header or socket connection

- Add logic to check for X-Forwarded-For header in response
- If header is not present, use socket connection to get host IP
- Update self.host with the obtained IP address

re-imagined committed:
Merge branch 'main' into vllm_client_custom_url

re-imagined committed:
fix(vllm_client): enhance server connectivity and IP retrieval

re-imagined committed:
Merge branch 'main' into vllm_client_custom_url

re-imagined committed:
Merge branch 'main' into vllm_client_custom_url

re-imagined commented:
@qgallouedec 
hi, it's ready for review now

re-imagined committed:
Merge branch 'main' into vllm_client_custom_url

re-imagined commented on trl/extras/vllm_client.py:
```diff
@@ -119,11 +152,17 @@ def check_server(self, total_timeout: float = 0.0, retry_interval: float = 2.0):
                     ) from exc
             else:
                 if response.status_code == 200:
+                    if "X-Forwarded-For" in response.headers:
```
we still need the ip address when using base url

re-imagined committed:
Merge branch 'vllm_client_custom_url' of github.com:re-imagined/trl into vllm_client_custom_url

re-imagined commented on trl/extras/vllm_client.py:
```diff
@@ -119,11 +152,19 @@ def check_server(self, total_timeout: float = 0.0, retry_interval: float = 2.0):
                     ) from exc
             else:
                 if response.status_code == 200:
+                    if "X-Forwarded-For" in response.headers:
```
we still need the ip addrees when using base url
1. An HTTP proxy can add a header containing the backend IP so that the client can read it from the response. The most common header for passing client IPs is X-Forwarded-For
2. get the ip via a request

qgallouedec replied:
I've updated the host to be `self.host = socket.gethostbyname(parsed_url.hostname)` during the init. It's simpler. It works locally, but I'm curious to know if it works for you.

re-imagined committed:
Merge branch 'main' into vllm_client_custom_url

re-imagined committed:
Merge branch 'main' into vllm_client_custom_url

qgallouedec committed:
Merge branch 'main' into vllm_client_custom_url

qgallouedec committed:
update test

qgallouedec committed:
get host when url is parsed

qgallouedec committed:
format

qgallouedec reviewed (commented):
Sorry for the late review. I've updated your branch with a few minor things, plus an attempt to simplify the host recovering. Can you check that it works on your side?

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/trl/pr_3324). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

re-imagined commented:
> Sorry for the late review. I've updated your branch with a few minor things, plus an attempt to simplify the host recovering. Can you check that it works on your side?

thank you, I will test it today

re-imagined committed:
Merge branch 'main' into vllm_client_custom_url

re-imagined commented:
> Sorry for the late review. I've updated your branch with a few minor things, plus an attempt to simplify the host recovering. Can you check that it works on your side?

@qgallouedec 
I’ve tested the latest changes on my end, and everything works as expected. The simplification to the host recovery logic is a nice improvement—thanks for taking the time to refine it.

qgallouedec reviewed (approved):
LGTM, thanks!

qgallouedec committed:
Merge branch 'main' into vllm_client_custom_url
